cmd/helm/downloader/manager.go

```diff
@@ -0,0 +1,323 @@
+func (m *Manager) downloadAll(deps []*chartutil.Dependency) error {
+	repos, err := m.loadChartRepositories()
+	if err != nil {
+		return err
+	}
+
+	dl := ChartDownloader{
+		Out:      m.Out,
+		Verify:   m.Verify,
+		Keyring:  m.Keyring,
+		HelmHome: m.HelmHome,
+	}
+
+	fmt.Fprintf(m.Out, "Saving %d charts\n", len(deps))
+	for _, dep := range deps {
+		fmt.Fprintf(m.Out, "Downloading %s from repo %s\n", dep.Name, dep.Repository)
+
+		target := fmt.Sprintf("%s-%s", dep.Name, dep.Version)
+		churl, err := findChartURL(target, dep.Repository, repos)
+		if err != nil {
+			fmt.Fprintf(m.Out, "WARNING: %s (skipped)", err)
+			continue
+		}
+
+		dest := filepath.Join(m.ChartPath, "charts")
+		if _, err := dl.DownloadTo(churl, dest); err != nil {
+			fmt.Fprintf(m.Out, "WARNING: Could not download %s: %s (skipped)", churl, err)
+			continue
+		}
+	}
+	return nil
+}
+
+// hasAllRepos ensures that all of the referenced deps are in the local repo cache.
+func (m *Manager) hasAllRepos(deps []*chartutil.Dependency) error {
+	rf, err := repo.LoadRepositoriesFile(m.HelmHome.RepositoryFile())
+	if err != nil {
+		return err
+	}
+	repos := rf.Repositories
+	// Verify that all repositories referenced in the deps are actually known
+	// by Helm.
+	missing := []string{}
+	for _, dd := range deps {
+		found := false
+		if dd.Repository == "" {
+			found = true
+		} else {
+			for _, repo := range repos {
+				if urlsAreEqual(repo, dd.Repository) {
+					found = true
+				}
+			}
+		}
+		if !found {
+			missing = append(missing, dd.Repository)
+		}
+	}
+	if len(missing) > 0 {
+		return fmt.Errorf("no repository definition for %s. Try 'helm repo add'", strings.Join(missing, ", "))
+	}
+	return nil
+}
+
+// UpdateRepositories updates all of the local repos to the latest.
+func (m *Manager) UpdateRepositories() error {
+	rf, err := repo.LoadRepositoriesFile(m.HelmHome.RepositoryFile())
+	if err != nil {
+		return err
+	}
+	repos := rf.Repositories
+	if len(repos) > 0 {
+		// This prints warnings straight to out.
+		m.parallelRepoUpdate(repos)
+	}
+	return nil
+}
+
+func (m *Manager) parallelRepoUpdate(repos map[string]string) {
+	out := m.Out
+	fmt.Fprintln(out, "Hang tight while we grab the latest from your chart repositories...")
+	var wg sync.WaitGroup
+	for name, url := range repos {
+		wg.Add(1)
+		go func(n, u string) {
+			err := repo.DownloadIndexFile(n, u, m.HelmHome.CacheIndex(n))
+			if err != nil {
+				updateErr := fmt.Sprintf("...Unable to get an update from the %q chart repository: %s", n, err)
+				fmt.Fprintln(out, updateErr)
+			} else {
+				fmt.Fprintf(out, "...Successfully got an update from the %q chart repository\n", n)
+			}
+			wg.Done()
+		}(name, url)
+	}
+	wg.Wait()
+	fmt.Fprintln(out, "Update Complete. Happy Helming!")
+}
+
+// urlsAreEqual normalizes two URLs and then compares for equality.
+func urlsAreEqual(a, b string) bool {
+	au, err := url.Parse(a)
+	if err != nil {
```

Comment:
Add comment: If two (identical) local paths are passed in, we treat those as a match.